src/include/transaction/transaction_context.h

```diff
@@ -203,6 +246,21 @@ class TransactionContext {
    */
   void SetMustAbort() { must_abort_ = true; }
 
+  storage::UndoBuffer GetUndoBuffer() { return undo_buffer_; }
+  storage::RedoBuffer GetRedoBuffer() { return redo_buffer_; }
+
+  void LockIfNotLocked(catalog::table_oid_t table_oid, common::ManagedPointer<std::shared_mutex> table_lock) {
+    if (!IsTableLocked(table_oid)) {
+      reinterpret_cast<DebugLock*>(table_lock.Get())->debug_lock_shared(common::ManagedPointer(this));
+      held_table_oids_.insert(table_oid);
```

Comment:
this looks like a race condition to me. For example, thread 1 can acquire the tablelock, and before thread 1 inserts to held_table_oids_, thread 2 checks IsTableLocked, which returns false. Then both threads get the lock

Reply:
IsTableLocked only applies to whether the current txn has the lock, so we don't double grab a lock. We are assuming that txns are single threaded wrt this function